extensions/amp-gfycat/0.1/amp-gfycat.js

```diff
@@ -48,35 +75,205 @@ class AmpGfycat extends AMP.BaseElement {
         'The data-gfyid attribute is required for <amp-gfycat> %s',
         this.element);
-    const noautoplay = this.element.hasAttribute('noautoplay');
+  }
 
-    const iframe = this.element.ownerDocument.createElement('iframe');
-    iframe.setAttribute('frameborder', '0');
+  /** @return {string} */
+  getVideoIframeSrc_() {
+    if (this.videoIframeSrc_) {
+      return this.videoIframeSrc_;
+    }
+
+    const videoid = this.videoid_ || '';
+    let src = 'https://gfycat.com/ifr/' + encodeURIComponent(videoid);
+
+    const params = getDataParamsFromAttributes(this.element);
 
-    let src = 'https://gfycat.com/ifr/' + encodeURIComponent(gfyid);
+    const noautoplay = this.element.hasAttribute('noautoplay');
     if (noautoplay) {
-      src += '?autoplay=0';
+      params['autoplay'] = '0';
     }
+    src = addParamsToUrl(src, params);
+    return this.videoIframeSrc_ = src;
+  }
 
+  /** @override */
+  layoutCallback() {
+    const iframe = this.element.ownerDocument.createElement('iframe');
+    const src = this.getVideoIframeSrc_();
+
+    iframe.setAttribute('frameborder', '0');
     iframe.src = src;
     this.applyFillContent(iframe);
-    this.element.appendChild(iframe);
     this.iframe_ = iframe;
 
-    return this.loadPromise(iframe);
+    this.unlistenMessage_ = listen(
+        this.win,
+        'message',
+        this.handleGfycatMessages_.bind(this)
+    );
+
+    this.element.appendChild(iframe);
+    const loaded = this.loadPromise(this.iframe_).then(() => {
+      this.element.dispatchCustomEvent(VideoEvents.LOAD);
+    });
+    this.playerReadyResolver_(loaded);
+    return loaded;
   }
 
   /** @override */
-  pauseCallback() {
-    if (this.iframe_ && this.iframe_.contentWindow) {
-      this.iframe_.contentWindow./*OK*/postMessage('pause', '*');
+  unlayoutCallback() {
+    if (this.iframe_) {
+      removeElement(this.iframe_);
+      this.iframe_ = null;
+    }
+    if (this.unlistenMessage_) {
+      this.unlistenMessage_();
+    }
+
+    this.playerReadyPromise_ = new Promise(resolve => {
+      this.playerReadyResolver_ = resolve;
+    });
+    return true;  // Call layoutCallback again.
+  }
+
+  /**
+   * Sends a command to the player through postMessage.
+   * @param {string} command
+   * @param {*=} opt_arg
+   * @private
+   * */
+  sendCommand_(command, opt_arg) {
+    this.playerReadyPromise_.then(() => {
+      if (this.iframe_ && this.iframe_.contentWindow) {
+        const message = command;
+        this.iframe_.contentWindow./*OK*/postMessage(message, '*');
+      }
+    });
+  }
+
+  /** @private */
+  handleGfycatMessages_(event) {
+    const eventData = /** @type {?string|undefined} */ (getData(event));
+
+    if (event.origin !== 'https://gfycat.com' ||
+        event.source != this.iframe_.contentWindow ||
+        typeof eventData !== 'string') {
+      return;
+    }
+
+    if (eventData == 'ready') {
+      this.element.dispatchCustomEvent(VideoEvents.LOAD);
```

Comment:
hmm, iframe load was also dispatching `VideoEvents.LOAD` only one is needed. For some players, iframe being loaded is still too early to be considered "ready" that's why they have this custom message. If that the case for you as well? Does this event come from the player at all? Pick between this or iframe load, we only need one.